miner/verify_bundles.go

```diff
@@ -0,0 +1,285 @@
+}
+
+func (e *ErrBundleTxWrongPlace) Error() string {
+	return fmt.Sprintf("tx from included bundle is in wrong place tx_hash=%s, bundle_hash=%s, tx_bundle_index=%d, tx_block_index=%d, expected_block_index=%d", e.TxHash.Hex(), e.BundleHash.Hex(), e.TxIndex, e.BlockIndex, e.ExpectedBlockIndex)
+}
+
+// ErrPrivateTxFromFailedBundle is returned when a private tx is included in the block, but the bundle it belongs to was not included
+type ErrPrivateTxFromFailedBundle struct {
+	BundleHash common.Hash
+	TxHash     common.Hash
+	// Index of the tx in the bundle
+	TxIndex int
+}
+
+func NewErrPrivateTxFromFailedBundle(bundleHash, txHash common.Hash, txIndex int) *ErrPrivateTxFromFailedBundle {
+	return &ErrPrivateTxFromFailedBundle{
+		BundleHash: bundleHash,
+		TxHash:     txHash,
+		TxIndex:    txIndex,
+	}
+}
+
+func (e *ErrPrivateTxFromFailedBundle) Error() string {
+	return fmt.Sprintf("private tx from failed bundle included in the block tx_hash=%s, bundle_hash=%s, tx_bundle_index=%d", e.TxHash.Hex(), e.BundleHash.Hex(), e.TxIndex)
+}
+
+// VerifyBundlesAtomicity checks that all txs from the included bundles are included in the block correctly
+// 1. We check that all non-reverted txs from the bundle are included in the block in correct order (with possible gaps) and are not reverted
+// 2. Reverted txs are allowed to be not included in the block
+// NOTE: we only verify bundles that were committed in the block but not all bundles that we tried to include
+func VerifyBundlesAtomicity(env *environment, committedBundles, allBundles []types.SimulatedBundle, usedSbundles []types.UsedSBundle, mempoolTxHashes map[common.Hash]struct{}) error {
+	// bundleHash -> tx
+	includedBundles := make(map[common.Hash][]bundleTxData)
+	extractBundleTxDataFromBundles(committedBundles, includedBundles)
+	extractBundleTxDataFromSbundles(usedSbundles, includedBundles, true)
+	includedTxDataByHash := extractIncludedTxDataFromEnv(env)
+
+	allUsedBundles := make(map[common.Hash][]bundleTxData)
+	extractBundleTxDataFromBundles(allBundles, allUsedBundles)
+	extractBundleTxDataFromSbundles(usedSbundles, allUsedBundles, false)
+	privateTxDataFromFailedBundles := extractPrivateTxsFromFailedBundles(includedBundles, allUsedBundles, mempoolTxHashes)
+
+	return checkBundlesAtomicity(includedBundles, includedTxDataByHash, privateTxDataFromFailedBundles)
+}
+
+type bundleTxData struct {
+	hash      common.Hash
+	canRevert bool
+}
+
+type includedTxData struct {
+	hash     common.Hash
+	index    int
+	reverted bool
+}
+
+type privateTxData struct {
+	bundleHash common.Hash
+	index      int
+}
+
+// checkBundlesAtomicity checks that all txs from the included bundles are included in the block correctly
+// 1. We check that all non-reverted txs from the bundle are included in the block and are not reverted
+// 2. Reverted txs are allowed to be not included in the block
+// 3. All txs from the bundle must be in the right order, gaps between txs are allowed
+func checkBundlesAtomicity(
```

Comment:
Would you maybe want to cross-check with the block itself?

Reply:
I use `environment` from `worker.go` to build input args to this function. Since `environment` is used to build block it should be as close to the block as possible. (we can't use `Block` because it does not have receipts inside)